mfr/extensions/zip/render.py

```diff
@@ -17,15 +16,56 @@ class ZipRenderer(extension.BaseRenderer):
 
     def render(self):
         zip_file = zipfile.ZipFile(self.file_path, 'r')
+        files = [file for file in zip_file.filelist if not file.filename.startswith('__MACOSX')]
 
-        filelist = [{'name': markupsafe.escape(file.filename),
-                     'size': sizeof_fmt(int(file.file_size)),
-                     'date': "%d-%02d-%02d %02d:%02d:%02d" % file.date_time[:6]} for file in zip_file.filelist
-                    if not file.filename.startswith('__MACOSX')]
+        data = self.filelist_to_tree(files)
 
-        message = '' if filelist else 'This zip file is empty.'
+        return self.TEMPLATE.render(data=data, base=self.assets_url)
 
-        return self.TEMPLATE.render(zipped_filenames=filelist, message=message)
+    def filelist_to_tree(self, files):
+
+        self.icons_url = self.assets_url + '/img/icons'
+
+        tree_data = [{
+            'text': self.metadata.name + self.metadata.ext,
+            'icon': self.icons_url + '/file_extension_zip.png',
+            'children': []
+        }]
+
+        for file in files:
+            node_path = tree_data[0]
+            paths = file.filename.split('/')
+            for i, path in enumerate(paths):
+                length = len(node_path['children'])
+                if not length or node_path['children'][length - 1]['text'] != path:
+
+                    new_node = dict(text=path)
+
+                    new_node['children'] = []
+
+                    if new_node['text']:
+                        date = '%d-%02d-%02d %02d:%02d:%02d' % file.date_time[:6]
+                        size = sizeof_fmt(int(file.file_size))
+                        new_node['data'] = {'date': date, 'size': size}
+
+                        if file.filename[-1] != '/':
+                            ext = os.path.splitext(file.filename)[1].lstrip('.')
+                            if check_icon_ext(ext):
+                                new_node['icon'] = \
+                                    self.icons_url + '/file_extension_{}.png'.format(ext)
+                            else:
+                                new_node[
+                                    'icon'] = self.icons_url + '/generic-file.png'
+                        else:
+                            new_node['icon'] = self.icons_url + '/folder.png'
+
+                        node_path['children'].append(new_node)
+
+                    node_path = new_node
+                else:
+                    node_path = node_path['children'][length - 1]
```

Comment:
Okay i understand the `if` statement a bit better, but there has to be a better way to write this.